### PR TITLE
Added doctrine_mongo_autocomplete model filter

### PR DESCRIPTION
**Pull request by joshlopes**
Q   |   A
------|-------
Bug fix? |	yes
New feature? |	yes
BC breaks? |	no
Deprecations? |	no
Tests pass? | yes
Fixed tickets | https://github.com/sonata-project/SonataDoctrineMongoDBAdminBundle/issues/138
License | no
Doc PR | no

### CODE REVIEW

**Comment by joshlopes** [comment]
Improved the DataGrid to look more similar to ORM (including the pager tag settings etc which for now we only have one Pager) 

For the autocomplete to work on multiple associations i had to also change the base Filter to look on parents association and clean it up

File: Builder/DatagridBuilder.php [hunk 1]
```diff
@@ -151,4 +168,28 @@ public function getBaseDatagrid(AdminInterface $admin, array $values = array())
...
+     *
+     * @throws \RuntimeException If invalid pager type is set.
+     */
+    protected function getPager($pagerType)
+    {
+        switch ($pagerType) {
+            case Pager::TYPE_DEFAULT:
+                return new Pager();
+
+            // TODO: Implement a Simple page that extends Simple Pager from admin?
```
Comment (rande) [inline Builder/DatagridBuilder.php]: Just remove the code ;)

Comment (joshlopes) [inline Builder/DatagridBuilder.php]: done

**Comment by rande** [comment]
@dunglas can you review this PR ?

**Comment by joshlopes** [comment]
Any news on this? @dunglas can you please review and tell me if anything is missing ? thanks

**Comment by joshlopes** [comment]
@rande any chance that you can review this yourself ? :)

**Comment by joshlopes** [comment]
Kind bump

**Comment by joshlopes** [comment]
@rande or @dunglas can you please check all Pulls and merge ? I have to walk with my vendors folder on my back to keep this code on my servers / projects. Would be great if you could merge / close this pull -- have been using them for 3 months now no problems whatsoever.

**Comment by dunglas** [comment]
@joshlopes can you resvole conflicts?

File: Builder/DatagridBuilder.php [hunk 2]
```diff
@@ -151,4 +168,23 @@ public function getBaseDatagrid(AdminInterface $admin, array $values = array())
...
+    /**
+     * Get pager by pagerType
+     *
+     * @param string $pagerType
+     *
+     * @return \Sonata\AdminBundle\Datagrid\PagerInterface
+     *
+     * @throws \RuntimeException If invalid pager type is set.
+     */
+    protected function getPager($pagerType)
```
Comment (dunglas) [inline Builder/DatagridBuilder.php]: What's the use case of this new extension point?

File: Filter/Filter.php [hunk 3]
```diff
@@ -28,6 +28,27 @@ public function apply($queryBuilder, $value)
...
     }
 
+    public function getFieldName()
+    {
+        $fieldName = $this->getOption('field_name');
+
+        if (is_array($this->getOption('parent_association_mappings'))) {
+            foreach($this->getOption('parent_association_mappings') as $map) {
+                if(!empty($map['name'])) {
+                    $fieldName = $map['name'] . "." . $fieldName;
```
Comment (dunglas) [inline Filter/Filter.php]: Please use simple quotes (run PHP CS Fixer).

File: Filter/Filter.php [hunk 4]
```diff
@@ -28,6 +28,27 @@ public function apply($queryBuilder, $value)
         $this->filter($queryBuilder, null, $this->getFieldName(), $value);
     }
 
+    public function getFieldName()
```
Comment (dunglas) [inline Filter/Filter.php]: Can you add some PHPDoc please?

File: Filter/ModelAutocompleteFilter.php [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+ * file that was distributed with this source code.
+ */
+
+namespace Sonata\DoctrineMongoDBAdminBundle\Filter;
+
+use Sonata\CoreBundle\Form\Type\EqualType;
+use Sonata\AdminBundle\Datagrid\ProxyQueryInterface;
+use Doctrine\Common\Collections\Collection;
+
+class ModelAutocompleteFilter extends Filter
```
Comment (dunglas) [inline Filter/ModelAutocompleteFilter.php]: Can you add some PHPDoc?

File: Filter/ModelAutocompleteFilter.php [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+{
+
+    public function filter(ProxyQueryInterface $queryBuilder, $alias, $field, $data)
+    {
+        if (!$data || !is_array($data) || !array_key_exists('value', $data)) {
```
Comment (dunglas) [inline Filter/ModelAutocompleteFilter.php]: Can be simplified by `if (!isset($data['value'])) {`

File: Filter/ModelAutocompleteFilter.php [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+            $this->handleMultiple($queryBuilder, $alias, $field, $data);
+        } else {
+            $this->handleScalar($queryBuilder, $alias, $field, $data);
+        }
+
+    }
+
+    /**
+     *
+     * @param  ProxyQueryInterface $queryBuilder
```
Comment (dunglas) [inline Filter/ModelAutocompleteFilter.php]: Please fix PHPDoc.

File: Filter/ModelAutocompleteFilter.php [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+     * @param  type                $alias
+     * @param  type                $field
+     * @param  type                $data
+     * @return type
+     */
+    protected function handleMultiple(ProxyQueryInterface $queryBuilder, $alias, $field, $data)
+    {
+        if (count($data['value']) == 0) {
```
Comment (dunglas) [inline Filter/ModelAutocompleteFilter.php]: Always use strict operators (`===`) and yoda style (`0 === count($data['value']))`.

File: Filter/ModelAutocompleteFilter.php [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+            return;
+        }
+
+        $ids = array();
+        foreach ($data['value'] as $value) {
+            $ids[] = self::fixIdentifier($value->getId());
+        }
+
+        if (isset($data['type']) && $data['type'] == EqualType::TYPE_IS_NOT_EQUAL) {
```
Comment (dunglas) [inline Filter/ModelAutocompleteFilter.php]: Same here (`=== `)

File: Filter/ModelAutocompleteFilter.php [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+            $queryBuilder->field($field)->notIn($ids);
+        } else {
+            $queryBuilder->field($field)->in($ids);
+        }
+
+        $this->active = true;
+    }
+
+    /**
+     *
```
Comment (dunglas) [inline Filter/ModelAutocompleteFilter.php]: Correct the PHPDoc.

File: Filter/ModelAutocompleteFilter.php [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+     * @param  ProxyQueryInterface $queryBuilder
+     * @param  type                $alias
+     * @param  type                $field
+     * @param  type                $data
+     * @return type
+     */
+    protected function handleScalar(ProxyQueryInterface $queryBuilder, $alias, $field, $data)
```
Comment (dunglas) [inline Filter/ModelAutocompleteFilter.php]: data should be typehinted `array $data`

File: Filter/ModelAutocompleteFilter.php [hunk 12]
```diff
@@ -0,0 +1,162 @@
...
+
+    /**
+     *
+     * @param  ProxyQueryInterface $queryBuilder
+     * @param  type                $alias
+     * @param  type                $field
+     * @param  type                $data
+     * @return type
+     */
+    protected function handleMultiple(ProxyQueryInterface $queryBuilder, $alias, $field, $data)
```
Comment (dunglas) [inline Filter/ModelAutocompleteFilter.php]: `array $data` (typehint)

File: Filter/ModelAutocompleteFilter.php [hunk 13]
```diff
@@ -0,0 +1,162 @@
...
+     */
+    protected function handleScalar(ProxyQueryInterface $queryBuilder, $alias, $field, $data)
+    {
+        if (empty($data['value'])) {
+            return;
+        }
+
+        $id = self::fixIdentifier($data['value']->getId());
+
+        if (isset($data['type']) && $data['type'] == EqualType::TYPE_IS_NOT_EQUAL) {
```
Comment (dunglas) [inline Filter/ModelAutocompleteFilter.php]: `===`

File: Filter/ModelAutocompleteFilter.php [hunk 14]
```diff
@@ -0,0 +1,162 @@
...
+            $queryBuilder->field($field)->equals($id);
+        }
+
+        $this->active = true;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function association(ProxyQueryInterface $queryBuilder, $data)
```
Comment (dunglas) [inline Filter/ModelAutocompleteFilter.php]: Is this extension point necessary? Make the function `private` if it isn't.

Comment (dunglas) [inline Filter/ModelAutocompleteFilter.php]: Don't mind my previous comment. It's inherited.

File: Filter/ModelAutocompleteFilter.php [hunk 15]
```diff
@@ -0,0 +1,162 @@
...
+        return array($alias, false);
+    }
+
+    /**
+     * Return \MongoId if $id is MongoId in string representation, otherwise custom string
+     *
+     * @param  mixed           $id
+     * @return \MongoId|string
+     */
+    protected static function fixIdentifier($id)
```
Comment (dunglas) [inline Filter/ModelAutocompleteFilter.php]: Why `static`?

File: Filter/ModelFilter.php [hunk 16]
```diff
@@ -18,13 +18,7 @@
 
 class ModelFilter extends Filter
 {
-    /**
```
Comment (dunglas) [inline Filter/ModelFilter.php]: Why do you delete this docblock?

File: Filter/ModelFilter.php [hunk 17]
```diff
@@ -18,13 +18,7 @@
...
-     * @param ProxyQueryInterface $queryBuilder
-     * @param string              $alias
-     * @param string              $field
-     * @param mixed               $data
-     * @return
-     */
+
     public function filter(ProxyQueryInterface $queryBuilder, $alias, $field, $data)
     {
         if (!$data || !is_array($data) || !array_key_exists('value', $data)) {
```
Comment (dunglas) [inline Filter/ModelFilter.php]: `isset`.

**Comment by dunglas** [comment]
Can you add tests for the new code?

File: Filter/Filter.php (hunk 3 above)
Comment (joshlopes) [inline Filter/Filter.php]: this was before the PHP CS Fixer

File: Builder/DatagridBuilder.php (hunk 2 above)
Comment (joshlopes) [inline Builder/DatagridBuilder.php]: I tried to keep the same logic of MySQL PHP file for be easier to mantain both projects by just comparing them -- till common areas such this one become merged into Admin bundle

Comment (dunglas) [inline Builder/DatagridBuilder.php]: :+1:

**Comment by joshlopes** [comment]
This is wrong, i am creating a new branch for this feature and use the master on my project as this review / test etc is taking to much time and all my work is doubled because there is major bugs that needs to be fixed on live to use this project. So -- my master branch will be the up-to-date with all features and the other minor branchs will be inserted here for this long review / test thingy..

If you are reading this and you are interested in having the features use my fork version instead.

Thanks!

**Pull request closed without merging**